content/recipes/async-local-storage.md

```diff
@@ -0,0 +1,228 @@
+  }
+}
+```
+
+4. That's it. Now we have a way to share request related state without needing to inject the whole `REQUEST` object.
+
+> warning **warning** Please be aware that while the technique is useful for many use-cases, it inherently obfuscates the code flow (creating implicit context), so use it responsibly and especially avoid creating contextual "[God objects](https://en.wikipedia.org/wiki/God_object)".
+
+### NestJS CLS
+
+The [nestjs-cls](https://github.com/Papooch/nestjs-cls) package provides several DX improvements over using plain `AsyncLocalStorage` (`CLS` is an abbreviation of the term _continuation-local storage_). It abstracts the implementation into a `ClsModule` that offers various ways of initializing the `store` for different transports (not only HTTP), as well as a strong-typing support.
+
+The store can then be accessed with an injectable `ClsService`, or entirely abstracted away from the business logic by using [Proxy Providers](https://www.npmjs.com/package/nestjs-cls#proxy-providers).
+
+> info **info** `nestjs-cls` is a third party package and is not managed by the NestJS core team. Please, report any issues found with the library in the [appropriate repository](https://github.com/Papooch/nestjs-cls/issues).
+
+#### Installation
+
+Apart from a peer dependency on the `@nestjs` libs, it only depends on the built-in Node.js API. Install it as any other package.
+
+```bash
+npm i nestjs-cls
+```
+
+#### Usage
+
+A similar functionality as described [above](#custom-implementation) can be implemented using `nestjs-cls` as follows:
+
+1. Import the `ClsModule` in the root module.
+
+```ts
+/** app.module.ts */
+
+@Module({
+  imports: [
+    // Register the ClsModule,
+    ClsModule.forRoot({
+      middleware: {
+        // automatically mount the
+        // ClsMiddleware for all routes
+        mount: true,
+        // and use the setup method to
+        // provide default store values.
+        setup: (cls, req) => {
+          cls.set('userId', req.headers['x-user-id']);
+        },
+      },
+    }),
+  ],
+  providers: [CatService],
+  controllers: [CatController],
+})
+export class AppModule {}
+```
+
+2. And then can use the `ClsService` to access the store values.
+
+```ts
+/** cat.service.ts */
+
+export class CatService {
+  constructor(
+    // We can inject the provided ClsService instance,
+    private readonly cls: ClsService,
+    private readonly catRepository: CatRepository,
+  ) {}
+
+  getCatForUser() {
+    // and use the "get" method to retrieve any stored value.
+    const userId = this.cls.get('userId');
+    return this.catRepository.getForUser(userId);
+  }
+}
+```
+
+3. To get strong typing of the store values managed by the `ClsService`, we can use an optional type parameter (`ClsService<MyStoreType>`), or use application-wide typescript module augmentation.
+
+```ts
+declare module `nestjs-cls` {
+  interface ClsStore {
+    userId: number
+  }
+}
+```
```

Comment:
nit: If there's a way to use a custom interface over module augmentation, I'd prefer we show that.